test/distributed/pipelining/test_schedule.py

```diff
@@ -497,6 +497,125 @@ def test_non_symmetric_stage_ids(self, ScheduleClass):
+    @skip_but_pass_in_sandcastle_if(not TEST_MULTIGPU, "NCCL test requires 2+ GPUs")
+    @parametrize("ScheduleClass", [ScheduleFlexibleInterleaved1F1B])
+    def test_schedule_with_weight_update_mlp_e2e(self, ScheduleClass):
+        return
+        stages_per_rank = 2
+        n_stages = stages_per_rank * self.world_size
+        full_mod = MultiMLPWithDw(d_hid, n_layers=n_stages)
+        full_mod.to(self.device)
+
+        ref_mod = copy.deepcopy(full_mod)
+        x = torch.randn(batch_size, d_hid, device=self.device)
+        with torch.no_grad():
+            y = ref_mod(x)
+            # Add a small perturbation
+            target = y + torch.randn(batch_size, d_hid, device=self.device)
+
+        ref_loss_fn = torch.nn.MSELoss(reduction="sum")
+        full_loss_fn = torch.nn.MSELoss(reduction="sum")
+
+        ref_optimizer = torch.optim.Adam(ref_mod.parameters(), lr=0.001)
+        full_optimizer = torch.optim.Adam(full_mod.parameters(), lr=0.001)
+
+        full_mod.toggle()
+
+        # Get a submodule, e.g. `layers.0` or `layers.1`
+        stage_indices = [
+            self.rank + i * self.world_size for i in range(stages_per_rank)
+        ]
+        submod_names = [f"layers.{i}" for i in stage_indices]
+        stage_modules = [
+            full_mod.get_submodule(submod_name) for submod_name in submod_names
+        ]
+
+        # Run reference
+        for _ in range(2):
+            ref_stage_modules = [
+                ref_mod.get_submodule(submod_name) for submod_name in submod_names
+            ]
+            for stage_module in ref_stage_modules:
+                stage_module.zero_grad()
+
+            ref_mod.zero_grad()
+            ref_out = ref_mod(x)
+            ref_loss = ref_loss_fn(ref_out, target)
+            ref_loss.backward()
+
+
+        class CustomState:
+            def __init__(self, stage_module, stage_idx, rank):
+                self.i = 0
+                self.stage_module = stage_module
+                self.stage_idx = stage_idx
+                self.rank = rank
+
+            def dw_builder(self):
+                def dw_runner():
+                    # This inner function would be called by PipelineStage during `backward_weight_one_chunk`
+                    self.i += 1
+                    print(f"[Rank {self.rank}] dw_count={self.i} stage={self.stage_idx}")
+                    self.stage_module.compute_dW()
+
+                return dw_runner
+
+        cs = {}
+        for stage_module, stage_idx in zip(stage_modules, stage_indices):
+            cs[stage_idx] = CustomState(stage_module, stage_idx, self.rank)
+
+        # Create a pipeline stage to wrap that submodule
+        chunks = 2
+        input_args = x.chunk(chunks)[0]
+        stages = [
+            PipelineStage(
+                stage_module,
+                stage_idx,
+                n_stages,
+                self.device,
+                input_args=input_args,
+                dw_builder=cs[stage_idx].dw_builder,
```

Comment:
we can probably simplify this a bit.  I initially put 'dw_runner' as the argument to PipelineStage, but then i realized that you might need a different 'dw_runner' after each call to forward/DX becuase DX produced new internal states for next dw.  

However, your implementation already hides the caching of internal states from dx inside the single 'dw_runner' so you don't actually make a new dw_runner on each step when dw_builder is called.

If you prefer, I can just kill the 'dw_builder' concept and you can pass your dw_runner directly into PipelineStage as `dw_builder = stage_module.compute_dW` 

note: I think CustomState class is not adding any additional value, it just helps print a log. I can probably log stage_idx when calling dw_runner inside PP code so that log is available to you.